hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/slowlog/SlowLogEventHandler.java

```diff
@@ -0,0 +1,205 @@
+    final String methodDescriptorName =
+      methodDescriptor != null ? methodDescriptor.getName() : StringUtils.EMPTY;
+    SlowLogPayload slowLogPayload = SlowLogPayload.newBuilder()
+      .setCallDetails(methodDescriptorName + "(" + param.getClass().getName() + ")")
+      .setClientAddress(clientAddress)
+      .setMethodName(methodDescriptorName)
+      .setMultiGets(numGets)
+      .setMultiMutations(numMutations)
+      .setMultiServiceCalls(numServiceCalls)
+      .setParam(slowLogParams != null ? slowLogParams.getParams() : StringUtils.EMPTY)
+      .setProcessingTime(processingTime)
+      .setQueueTime(qTime)
+      .setRegionName(slowLogParams != null ? slowLogParams.getRegionName() : StringUtils.EMPTY)
+      .setResponseSize(responseSize)
+      .setServerClass(className)
+      .setStartTime(startTime)
+      .setUserName(userName)
+      .build();
+    queue.add(slowLogPayload);
+  }
+
+  /**
+   * Cleans up slow log payloads
+   *
+   * @return true if slow log payloads are cleaned up, false otherwise
+   */
+  boolean clearSlowLogs() {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Received request to clean up online slowlog buffer..");
+    }
+    queue.clear();
+    return queue.isEmpty();
+  }
+
+  /**
+   * Retrieve list of slow log payloads
+   *
+   * @param request slow log request parameters
+   * @return list of slow log payloads
+   */
+  List<SlowLogPayload> getSlowLogPayloads(final AdminProtos.SlowLogResponseRequest request) {
+    List<SlowLogPayload> slowLogPayloadList =
+      Arrays.stream(queue.toArray(new SlowLogPayload[0])).collect(Collectors.toList());
+    Collections.reverse(slowLogPayloadList);
+    if (isFilterProvided(request)) {
+      slowLogPayloadList = filterSlowLogs(request, slowLogPayloadList);
+    }
+    int limit = request.getLimit() >= slowLogPayloadList.size() ? slowLogPayloadList.size()
+      : request.getLimit();
+    return slowLogPayloadList.subList(0, limit);
+  }
+
+  private boolean isFilterProvided(AdminProtos.SlowLogResponseRequest request) {
+    if (request.hasUserName() && StringUtils.isNotEmpty(request.getUserName())) {
+      return true;
+    }
+    if (request.hasTableName() && StringUtils.isNotEmpty(request.getTableName())) {
+      return true;
+    }
+    if (request.hasClientAddress() && StringUtils.isNotEmpty(request.getClientAddress())) {
+      return true;
+    }
+    return request.hasRegionName() && StringUtils.isNotEmpty(request.getRegionName());
+  }
+
+  private List<SlowLogPayload> filterSlowLogs(AdminProtos.SlowLogResponseRequest request,
+      List<SlowLogPayload> slowLogPayloadList) {
+    List<SlowLogPayload> filteredSlowLogPayloads = new ArrayList<>();
+    for (SlowLogPayload slowLogPayload : slowLogPayloadList) {
+      if (request.hasRegionName() && StringUtils.isNotEmpty(request.getRegionName())) {
+        if (slowLogPayload.getRegionName().equals(request.getRegionName())) {
+          filteredSlowLogPayloads.add(slowLogPayload);
```

Comment:
Actually this is counter intuitive from an operator POV. In the example you gave  ( {'REGION_NAME' => 'abc.12345.xxx.', 'TABLE_NAME' => 'def'}), it makes sense because region names are unique.. however

> Again for region_name="foo" and client="ip" && user="viraj", user would be interested to know either of them match not all of them because for slow_responses,  we want to retrieve as many entries available online as possible, rather than combination of filters.

Hmmm.. don't agree.. but it is what it is.. Either way this needs to be documented..like I said, this is counter intuitive .. so doesn't hurt to add a line about this in the docs.

Reply:
This is documented here already:
```
  hbase> get_slowlog_responses '*', {'REGION_NAME' => 'hbase:meta,,1', 'TABLE_NAME' => 't1'}
                                                                   => get slowlog responses with given region name
                                                                      or table name
  hbase> get_slowlog_responses '*', {'USER' => 'user_name', 'CLIENT_IP' => '192.162.1.40:60225'}
                                                                   => get slowlog responses that match either
                                                                      provided client IP address or user name

```